### ISSUE TITLE
Mobile_sam onnx inference

### ISSUE THREAD

**Issue by kyz-xxl** [ChaoningZhang/MobileSAM#170]
I've done the onnx export of Mobilesam, how should I reason about it?

**Comment by BlankHead2004** [comment]
Can you please share the file with me 
it would be really helpful

**Comment by kyz-xxl** [comment]
def export_encoder_model(model_type, checkpoint, output):
    """
    导出图像编码器为 ONNX 模型。

    Args:
        model_type (str): 模型类型。
        checkpoint (str): 模型权重文件路径。
        output (str): 输出 ONNX 文件路径。
    """
    import torch
    from mobile_sam import sam_model_registry

    device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
    sam = sam_model_registry[model_type](checkpoint=checkpoint)
    sam.to(device)
    sam.eval()

    # 使用批量大小为 1 的 dummy_image
    dummy_image = torch.randn(1, 3, 1024, 1024).to(device)
    dynamic_axes = {
        "input_image": {0: "batch_size", 2: "height", 3: "width"},
        "image_embeddings": {0: "batch_size"}  # 确保输出张量的动态维度
    }
    output_names = ["image_embeddings"]

    torch.onnx.export(
        sam.image_encoder,
        dummy_image,
        output,
        export_params=True,
        opset_version=16,
        do_constant_folding=True,
        input_names=["input_image"],
        output_names=output_names,
        dynamic_axes=dynamic_axes
    )


def export_decoder_model(model_type, checkpoint, output):
    """
    导出掩码解码器为 ONNX 模型。

    Args:
        model_type (str): 模型类型。
        checkpoint (str): 模型权重文件路径。
        output (str): 输出 ONNX 文件路径。
    """
    import torch
    from mobile_sam import sam_model_registry
    from mobile_sam.utils.onnx import SamOnnxModel

    device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
    sam = sam_model_registry[model_type](checkpoint=checkpoint)
    sam.to(device)
    sam.eval()

    onnx_model = SamOnnxModel(sam, return_single_mask=False)

    dummy_inputs = {
        "image_embeddings": torch.randn(1, 256, 64, 64).to(device),
        "point_coords": torch.randint(low=0, high=1024, size=(1, 5, 2), dtype=torch.float).to(device),
        "point_labels": torch.randint(low=0, high=4, size=(1, 5), dtype=torch.float).to(device),
        "mask_input": torch.zeros(1, 1, 256, 256).to(device),
        "has_mask_input": torch.zeros(1, 1).to(device),
        "orig_im_size": torch.tensor([1024, 1024], dtype=torch.float32).to(device)
    }

    dynamic_axes = {
        "image_embeddings": {0: "batch_size"},
        "point_coords": {1: "num_points"},
        "point_labels": {1: "num_points"},
        "mask_input": {0: "batch_size"},
        "has_mask_input": {0: "batch_size"},
        "masks": {0: "batch_size"},
        "iou_predictions": {0: "batch_size"},
        "low_res_masks": {0: "batch_size"}
    }

    output_names = ["masks", "iou_predictions", "low_res_masks"]

    torch.onnx.export(
        onnx_model,
        tuple(dummy_inputs.values()),
        output,
        export_params=True,
        opset_version=16,
        do_constant_folding=True,
        input_names=list(dummy_inputs.keys()),
        output_names=output_names,
        dynamic_axes=dynamic_axes
    )


import argparse

def run_export():
    # parser = argparse.ArgumentParser(description="Export MobileSAM models to ONNX format.")
    # parser.add_argument("--model-type",default='vit_t', type=str, required=True, help="Type of the model (e.g., vit_t).")
    # parser.add_argument("--checkpoint", default='./weights/mobile_sam.pt', type=str, required=True, help="Path to the model checkpoint.")
    # parser.add_argument("--output-encoder", default='./weights/mobile_sam_encoder.onnx', type=str, required=True, help="Output path for the encoder ONNX model.")
    # parser.add_argument("--output-decoder", default='./weights/mobile_sam_decoder.onnx', type=str, required=True, help="Output path for the decoder ONNX model.")
    # args = parser.parse_args()

    # 导出图像编码器
    export_encoder_model('vit_t', '../weights/mobile_sam.pt', '../weights/mobile_sam_encoder.onnx')
    print(f"Encoder model successfully exported to ../weights/mobile_sam_encoder.onnx")

    # # 导出掩码解码器
    export_decoder_model('vit_t', '../weights/mobile_sam.pt', '../weights/mobile_sam_decoder.onnx')
    print(f"Decoder model successfully exported to ../weights/mobile_sam_decoder.onnx")

if __name__ == "__main__":
    run_export()



This is the conversion code I use, and you can also use the one provided officially.
